File: 이분 탐색 부수기/BOJ_28449_누가_이길까.java
package BOJ_28449;

import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BOJ_28449_누가_이길까 {

    // https://www.acmicpc.net/problem/28449
    // input
    private static BufferedReader br;

    // variables
    private static int N, M;
    private static int[] hiArr, arcArr;

    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("C:\\Users\\bigyo\\Desktop\\알고리즘\\JavaAlgorithm\\src\\BOJ_28449\\res.txt"));
        br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        input();

        bw.write(solve());
        bw.close();
    } // End of main()

    private static String solve() {
        StringBuilder sb = new StringBuilder();

        long draw = 0;
        long hiWin = 0;
        long arcWin = 0;
        for (int i = 0; i < M; i++) {
            int lower = lowerBound(0, N, arcArr[i]);
            int upper = upperBound(0, N, arcArr[i]);

            int temp = upper - lower;
            draw += temp;
            hiWin += N - upper;
            arcWin += upper - temp;
        }

        sb.append(hiWin).append(' ').append(arcWin).append(' ').append(draw);
        return sb.toString();
    } // End of solve()

    private static int lowerBound(int low, int high, int target) {
        if (low == high) return low;

        int mid = (low + high) / 2;
        if (hiArr[mid] < target) {
            return lowerBound(mid + 1, high, target);
        } else {
            return lowerBound(low, mid, target);
        }
    } // End of lowerBound()

    private static int upperBound(int low, int high, int target) {
        if (low == high) return low;

        int mid = (low + high) / 2;
        if (hiArr[mid] <= target) {
            return upperBound(mid + 1, high, target);
        } else {
            return upperBound(low, mid, target);
        }
    } // End of upperBound()

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        hiArr = new int[N];
        arcArr = new int[M];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            hiArr[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < M; i++) {
            arcArr[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(hiArr);
    } // End of input()
} // End of Main class

File: 이분 탐색 부수기/BOJ 28449 누가 이길까.kt
package BOJ_28449

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.File
import java.io.OutputStreamWriter
import java.util.*

/*
    이분탐색을 통해서 각 값의 중앙값을 찾는다.

    그리고 중앙을 기준으로 낮은 값들은 내가 다 이기는 값이 되고,
    높은 값들은 지는 값들이 된다.

 */


// https://www.acmicpc.net/problem/28449
// input
private lateinit var br: BufferedReader

// variables
private var N = 0
private var M = 0
private lateinit var HIArr: IntArray
private lateinit var ARCArr: IntArray

fun main() {
    val path = "C:\\Users\\bigyo\\Desktop\\알고리즘\\KotlinAlgo\\src\\main\\kotlin\\BOJ_28449\\res.txt"
    br = BufferedReader(File(path).bufferedReader())
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    input()

    bw.write(solve())
    bw.close()
} // End of main()

private fun solve(): String {
    val sb = StringBuilder()
    var hiWin = 0L
    var draw = 0L
    var arcWin = 0L

    for (i in 0 until M) {
        val lower = lowerBound(0, N, ARCArr[i])
        val upper = upperBound(0, N, ARCArr[i])
        val d = upper - lower

        draw += d
        hiWin += N - upper
        arcWin += upper - d
    }

    sb.append(hiWin).append(' ').append(arcWin).append(' ').append(draw)
    return sb.toString()
} // End of solve()

private fun lowerBound(low: Int, high: Int, target: Int): Int {
    if (low == high) return low

    val mid = (low + high) / 2
    return if (HIArr[mid] < target) {
        lowerBound(mid + 1, high, target)
    } else lowerBound(low, mid, target)
} // End of lowerBound()

// 찾고자 하는 값 이하의 값 중 가장 큰 값을 갖는 요소의 인덱스 + 1 을 반환
fun upperBound(low: Int, high: Int, target: Int): Int {
    if (low == high) return low

    val mid = (low + high) / 2
    return if (HIArr[mid] <= target) {
        upperBound(mid + 1, high, target)
    } else upperBound(low, mid, target)
} // End of upperBound()

private fun input() {
    StringTokenizer(br.readLine()).run {
        N = nextToken().toInt()
        M = nextToken().toInt()
    }

    StringTokenizer(br.readLine()).run {
        HIArr = IntArray(N) {
            nextToken().toInt()
        }
    }

    StringTokenizer(br.readLine()).run {
        ARCArr = IntArray(M) {
            nextToken().toInt()
        }
    }

    HIArr.sort()
} // End of input()
